working/macros/api/macros.dart

```diff
@@ -0,0 +1,82 @@
+import 'dart:async';
+
+import 'builders.dart';
+import 'introspection.dart';
+
+/// The marker interface for all types of macros.
+abstract class Macro {
+  const Macro();
+
+  external void buildTypes(FutureOr<void> Function(TypeBuilder) callback);
```

Comment:
`external`? Why not just abstract?

Reply:
A few reasons:

- It makes it so that the example macros are valid without actually implementing these.
- I actually do expect the implementations to be external (filled in with patch files). They probably won't be identical across analyzer/cfe/etc.

Reply:
I could instead hang these types off of another interface, and pass a 2nd argument to the `visit` methods which is of that type.

Reply:
Ah, I get it. So this is a concrete method that macro implementations can call. The question then is where to put that method. I think there's basically two options:

1. Put it on a class that we pass an instance of to the macro when it runs. Sort of how the builder API works.
2. Put it on the macro class itself so that the user's macro class can call it on itself.

If we're going to do the latter, then I think the natural way to model it is to make `Macro` be an abstract base class (or mixin) that concrete macro classes extend. Then `buildTypes()` can be a normal concrete instance method. If Dart had protected methods, it would probably be protected (and non-virtual).

I'm not sure which is the best approach, but if we do want this to be a method on the macro itself, then I wouldn't make it an `external` method on a `Macro` interface that gets magically implemented. That's pretty weird for users.

It's probably a little safer to take option (1) and hang it off a parameter that we pass to the macro.

Reply:
> It's probably a little safer to take option (1) and hang it off a parameter that we pass to the macro.

Ya, I agree. It would be very bizarre for somebody to override these methods haha.

I can work on moving these to other interfaces.

Reply:
Ok I did this - there are now a _lot_ of builder interfaces though. And the naming gets interesting (I ended up calling them all `builder` and shadowing local variables, to avoid verbosity in the names, but its gross).